node/network/collator-protocol/src/validator_side/mod.rs

```diff
@@ -887,165 +900,365 @@ async fn process_incoming_peer_message<Context>(
+	prospective_candidate: Option<(CandidateHash, Hash)>,
+) where
+	Sender: CollatorProtocolSenderTrait,
+{
+	let _span = state
+		.span_per_relay_parent
+		.get(&relay_parent)
+		.map(|s| s.child("advertise-collation"));
+
+	// First, perform validity checks:
+	// - Relay parent is known
+	// - Peer is declared
+	// - Para id is indeed the one we're assigned to at the given relay parent
+	// - Collator is not trying to build on top of occupied core (unless async
+	// backing is enabled)
+
+	let per_relay_parent = match state.per_relay_parent.get_mut(&relay_parent) {
+		Some(state) => state,
+		None => {
+			gum::debug!(
+				target: LOG_TARGET,
+				peer_id = ?peer_id,
+				?relay_parent,
+				"Advertise collation out of view",
+			);
+
+			modify_reputation(sender, *peer_id, COST_UNEXPECTED_MESSAGE).await;
+			return
+		},
+	};
+	let relay_parent_mode = per_relay_parent.prospective_parachains_mode;
+	let assignment = per_relay_parent.assignment;
+
+	let peer_data = match state.peer_data.get_mut(&peer_id) {
+		None => {
+			gum::debug!(
+				target: LOG_TARGET,
+				peer_id = ?peer_id,
+				?relay_parent,
+				"Advertise collation message has been received from an unknown peer",
+			);
+			modify_reputation(sender, *peer_id, COST_UNEXPECTED_MESSAGE).await;
+			return
+		},
+		Some(p) => p,
+	};
+
+	let para_id = if let Some(id) = peer_data.collating_para() {
+		id
+	} else {
+		gum::debug!(
+			target: LOG_TARGET,
+			peer_id = ?peer_id,
+			?relay_parent,
+			"Advertise collation message received from undeclared peer",
+		);
+		modify_reputation(sender, *peer_id, COST_UNEXPECTED_MESSAGE).await;
+		return
+	};
+
+	let core_state = match assignment.current {
+		Some((id, core_state)) if id == para_id => core_state,
+		_ => {
+			gum::debug!(
+				target: LOG_TARGET,
+				peer_id = ?peer_id,
+				para_id = ?para_id,
+				?relay_parent,
+				"Advertise collation message for para we're no assigned to",
+			);
+			modify_reputation(sender, *peer_id, COST_UNEXPECTED_MESSAGE).await;
+			return
+		},
+	};
+
+	if !relay_parent_mode.is_enabled() && core_state.is_occupied() {
+		gum::debug!(
+			target: LOG_TARGET,
+			peer_id = ?peer_id,
+			?relay_parent,
+			"Advertise collation message for an occupied core (async backing disabled)",
+		);
+		modify_reputation(sender, *peer_id, COST_UNEXPECTED_MESSAGE).await;
+		return
+	}
+
+	let insert_result = match (relay_parent_mode, prospective_candidate) {
+		(ProspectiveParachainsMode::Disabled, None) => peer_data.insert_advertisement(
+			relay_parent,
+			relay_parent_mode,
+			None,
+			&state.implicit_view,
+			&state.active_leaves,
+		),
+		(ProspectiveParachainsMode::Enabled, Some((candidate_hash, parent_head_data_hash))) => {
+			// TODO [now]: request hypothetical depth and check for backed parent nodes
+			// in a fragment tree.
+			peer_data.insert_advertisement(
+				relay_parent,
+				relay_parent_mode,
+				Some(candidate_hash),
+				&state.implicit_view,
+				&state.active_leaves,
+			)
+		},
```

Comment:
Given https://github.com/paritytech/polkadot/pull/5824 hypothetical depth request will return `(depth, ParentNodeCandidateHash, is_parent_backed)` **for each leaf in our view**

If for some leaf there exists a backed parent node, we're good to go fetch and validate collation.
Options are:
1) Discard advertisement otherwise
2) Keep an advertisement until we get notified about backed parent node
2.1 What to do if leaf gets deactivated? Hypothetical depth only makes sense with respect to some leaf, we may re-request it for descendant leaf and keep it around until we do this `ANCESTRY_LEN` times, but this looks bad to me.

Reply:
TBD in https://github.com/paritytech/polkadot/issues/5923